### ISSUE TITLE
all: test failures on netbsd-386-9_3 blocked in `os.wait6`

### ISSUE THREAD

**Issue by bcmills** [golang/go#57999]
```
#!watchflakes
post <- builder == "netbsd-386-9_3" && `^\s*os\.wait6`
```

**Comment by bcmills** [comment]
(attn @golang/netbsd)

See previously:
- #56180
- #50138
- #56215
- #48789

**Comment by bcmills** [comment]
@golang/netbsd, is anyone looking into this? These failures are quite frequent.

**Comment by bcmills** [comment]
@golang/netbsd, this failure rate is still quite high compared to the other ports. Is anyone looking into it?

**Comment by bcmills** [comment]
@golang/netbsd, the [Go porting policy](https://go.dev/wiki/PortingPolicy#requirements-for-a-new-port) requires that each port must have at least two developers “making required updates in a timely manner”.

Since no one has even responded to this issue since it was reported in January, and given the frequency of these failures, I believe that this port is no longer meeting that requirement. (If I am mistaken and someone is actually looking into this problem, please comment here with some details about when we can expect progress on a fix.)

**Comment by ianlancetaylor** [comment]
I didn't check every log but each log that I checked is consistent with a hang in the `wait6` system call.  I would vote for rolling back CL 431855 (again).  CC @tklauser

**Comment by bsiegert** [comment]
Would it help if we updated all the NetBSD 9.x test machines to the latest NetBSD 9_STABLE? That should have the rtld fix.

(Relatedly, I have a bug assigned to me to provide NetBSD 10 build VMs.)

**Comment by bcmills** [comment]
The Go bug behind #50138 was fixed in https://go.dev/cl/442478. It is possible that related bugs remain, but the cause of the hands must be different from that cause.

The libc bug behind #56810 was supposed to have been fixed by upgrading the builders in #54773.

**Comment by ianlancetaylor** [comment]
I don't think that history is quite right.  We don't have to implement `blockUntilWaitable` at all.  For NetBSD our first attempt at implementing `blockUntilWaitable` led to #48789.  We fixed that issue and tried again.  I am suggesting that that exposed this issue.  I think we should now try not implementing `blockUntilWaitable` at all on NetBSD.

**Comment by ianlancetaylor** [comment]
To be clear, I am suggesting that this is a NetBSD kernel bug in `wait6`.

**Comment by ianlancetaylor** [comment]
> 1. Just to be sure, can you share the output of `ident /usr/lib/libpthread.so` (or whatever is the pthread library that golang links against) on the affected builders?

Note that Go by default does not link against libpthread.so at all.  It calls the system call `lwp_create` directly.  That said, there are cases where it calls `pthread_create`, and I haven't checked which case applies for these failures.  The output of the `ident` is:

<details>

```
/usr/lib/libpthread.so:
     $NetBSD: crti.S,v 1.1 2010/08/07 18:01:35 joerg Exp $
     $NetBSD: crtbegin.S,v 1.2 2010/11/30 18:37:59 joerg Exp $
     $NetBSD: tss.c,v 1.1 2019/04/24 11:43:19 kamil Exp $
     $NetBSD: thrd.c,v 1.3.4.1 2019/09/13 06:54:09 martin Exp $
     $NetBSD: mtx.c,v 1.1 2019/04/24 11:43:19 kamil Exp $
     $NetBSD: cnd.c,v 1.1 2019/04/24 11:43:19 kamil Exp $
     $NetBSD: call_once.c,v 1.2 2019/04/24 21:41:15 kamil Exp $
     $NetBSD: sem.c,v 1.9 2019/02/21 21:33:34 christos Exp $
     $NetBSD: pthread_spin.c,v 1.6 2012/08/16 04:49:47 matt Exp $
     $NetBSD: pthread_once.c,v 1.3 2013/03/21 16:49:12 christos Exp $
     $NetBSD: pthread_cancelstub.c,v 1.38.32.1 2021/10/05 11:02:18 martin Exp $
     $NetBSD: pthread_tsd.c,v 1.17 2019/03/05 01:35:52 christos Exp $
     $NetBSD: pthread_specific.c,v 1.26 2013/03/21 16:49:12 christos Exp $
     $NetBSD: pthread_rwlock.c,v 1.34.18.1 2020/01/26 10:55:16 martin Exp $
     $NetBSD: pthread_mutex.c,v 1.65.2.1 2020/01/26 10:55:16 martin Exp $
     $NetBSD: pthread_misc.c,v 1.15.32.1 2020/01/26 10:55:16 martin Exp $
     $NetBSD: pthread_lock.c,v 1.34 2008/04/28 20:23:01 martin Exp $
     $NetBSD: pthread_cond.c,v 1.65.6.1 2020/01/26 10:55:16 martin Exp $
     $NetBSD: pthread_barrier.c,v 1.20 2016/07/03 14:24:58 christos Exp $
     $NetBSD: pthread_attr.c,v 1.18 2017/08/01 12:31:45 martin Exp $
     $NetBSD: pthread.c,v 1.153.2.2 2022/08/08 17:16:53 martin Exp $
     $NetBSD: pthread_makelwp_netbsd.c,v 1.2 2014/12/17 01:49:08 pooka Exp $
     $NetBSD: crtend.S,v 1.1 2010/08/07 18:01:34 joerg Exp $
     $NetBSD: crtn.S,v 1.1 2010/08/07 18:01:35 joerg Exp $
```

</details>

> 2. Is there a summary of how golang uses the wait family of system calls and handles SIGCHLD?  I see there's a lot of past discussion and changes, so it would be helpful if I had a clear idea of the current picture.

Go will normally install a signal handler for `SIGCHLD`, but for most programs, including the ones that are failing, that signal handler won't do anything.

When Go is waiting for a process, it first calls `wait6(P_PID, pidOfProcess, WEXITED|WNOWAIT)`.  When that call returns, Go will call `wait4(pidOfProcess, &status, 0, &rusage)`.  (This two step process is to avoid a race in sending a signal to a process and accidentally sending it to a PID of a process that has been waited for and then recycled.)

> 3. golang doesn't by any chance use SA_NOCLDWAIT, or set SIGCHLD to SIG_IGN, does it?

It does not.

> 4. Does golang have any mechanism to guarantee that it only calls waitid/wait6 without WNOWAIT at most once -- or, at most once with a successful return if WNOHANG -- for each child?  (I suppose the answer is probably yes but I want to rule out easy problems before going into a wild goose chase of harder ones!)

Go does not have any such mechanism.  If a program calls `os.Process.Wait` from two different goroutines simultaneously, then Go will call `wait4` from two different threads.  That said, that is not happening in these test cases.

> 5. Is there any ptrace involved?

No.

> 6. Is there any SIGTSTP/SIGSTOP/SIGCONT/WSTOPPED/WUNTRACED involved?

No.

> 7. Can you get any kind of core dump or stack trace of the child process when it hangs like this?

As far as I can tell, the child process has exited.

> 8. Are there any reports that consistently happen running the same test (after the rtld/pthread fix and the idtype numbering fix), so I could try running that one in a loop to more reliably reproduce it?

I haven't been able to recreate the problem.  It's fairly rare.  But I also haven't tried very hard.

> 9. Is there any chance you could start ktrace in the test harness, and save the ktrace file when it fails in this manner?  Something like `ktrace("/tmp/ktrace.out", KTROP_SET|KTRFLAG_DESCEND|KTRFACv2, KTRFAC_SYSCALL|KTRFAC_SYSRET|KTRFAC_NAMEI|KTRFAC_PSIG|KTRFAC_EMUL|KTRFAC_MIB, getpid())` when each test starts, and `ktrace(NULL, KTROP_CLEAR|KTRFLAG_DESCEND|KTRFACv2, KTRFAC_MASK, 1)` when it ends, and, if the test fails, either save /tmp/ktrace.out or include the output of `kdump -T -f /tmp/ktrace.out` in the report?  (@bsiegert, maybe you could rig this up?)

I'll leave that for @bsiegert.

> I took a look at a few other logs and they all appear to be blocking in wait6 under blockUntilWaitable, which is consistent with the hypothesis that the subprocess just hasn't exited yet. If that is the case, we need to know why the subprocess hasn't exited, not why wait6 isn't returning.

That is a fair point.  My only real reason for thinking that the problem is with `wait6` is that we aren't seeing the builder hang.  We're only seeing the builder time out while waiting for child processes.

**Comment by bcmills** [comment]
> I took a look at a few other logs and they all appear to be blocking in wait6 under blockUntilWaitable, which is consistent with the hypothesis that the subprocess just hasn't exited yet. If that is the case, we need to know why the subprocess hasn't exited, not why wait6 isn't returning.

Several of the failures in this log are for subtests of `os/exec.TestCancelErrors` — namely, subtests beginning with `killed_`.
In normal operation, those subtests terminate the child process by sending it `SIGKILL`.

For example, in https://build.golang.org/log/0889975bf55f90c65f04d4535e58591d7c30fa4f we see:
```
goroutine 172 [chan send, 29 minutes]:
runtime.gopark(0x40001d9df8?, 0xaf9f8?, 0x0?, 0x0?, 0x4000024000?)
	/var/gobuilder/buildlet/go/src/runtime/proc.go:398 +0xc8 fp=0x40001d9dd0 sp=0x40001d9db0 pc=0x4adf8
runtime.chansend(0x4000185560, 0x40001d9f90, 0x1, 0x40001d9e50?)
	/var/gobuilder/buildlet/go/src/runtime/chan.go:259 +0x3c4 fp=0x40001d9e40 sp=0x40001d9dd0 pc=0x166e4
runtime.chansend1(0x40001d9f10?, 0x3d4608?)
	/var/gobuilder/buildlet/go/src/runtime/chan.go:145 +0x18 fp=0x40001d9e70 sp=0x40001d9e40 pc=0x16308
os/exec.(*Cmd).watchCtx(0x4000235080, 0x4000185560)
	/var/gobuilder/buildlet/go/src/os/exec/exec.go:813 +0x328 fp=0x40001d9fb0 sp=0x40001d9e70 pc=0x10c368
os/exec.(*Cmd).Start.func4()
	/var/gobuilder/buildlet/go/src/os/exec/exec.go:715 +0x2c fp=0x40001d9fd0 sp=0x40001d9fb0 pc=0x10bdbc
runtime.goexit()
	/var/gobuilder/buildlet/go/src/runtime/asm_arm64.s:1197 +0x4 fp=0x40001d9fd0 sp=0x40001d9fd0 pc=0x7da24
created by os/exec.(*Cmd).Start in goroutine 120
	/var/gobuilder/buildlet/go/src/os/exec/exec.go:715 +0x788
```

That goroutine is blocked on this line:
https://cs.opensource.google/go/go/+/master:src/os/exec/exec.go;l=813;drc=29253f4d513f1d3fe549db39a8daa311492138b3
which is dominated by an unconditional call to `Kill` on the command's `*os.Process`:
https://cs.opensource.google/go/go/+/master:src/os/exec/exec.go;l=768;drc=29253f4d513f1d3fe549db39a8daa311492138b3

The implementation of `*os.Process.Kill` is pretty straightforward, so I would be pretty shocked if that doesn't end up making the correct syscalls on NetBSD. 😅

`os/exec.TestCancelErrors` and `os.TestStdPipe` might be good cases to focus on, because they involve only single subprocesses. (Many of the other failures here end up stuck on `go build`, but in those cases more often than not `go build` is itself blocked on a stuck invocation of `cmd/compile`, which makes the goroutine dump for the test itself look — correctly! — like it is blocked on a `go` subprocess that is still running.)

**Comment by bcmills** [comment]
> Is there a document anywhere that lays out how this works on each OS, or roughly how things like the syscalls and cgo dynamic symbol calls are put together?

I'm not aware of any particular document. (I would expect to find it in https://cs.opensource.google/go/go/+/master:src/runtime/HACKING.md, but it's not there.)

A short summary of my understanding:

- On `darwin` (including `ios`), `aix`, `solaris` (including `illumos`), `openbsd` (#61546 notwithstanding), and `windows`, the runtime always makes its system calls through a dynamically-linked `libc` or `kernel32.dll`.

- On `linux` (including `android`), `freebsd`, `netbsd`, `dragonfly`, `plan9`, and `wasm`, the runtime makes direct system calls using assembly routines.
   - However, if cgo is enabled and the program ends up linking against a package that uses it (often the `net` package), the runtime calls through libc for functions specific to thread management (declared in [`src/runtime/cgo.go`](https://cs.opensource.google/go/go/+/master:src/runtime/cgo.go;drc=c426c87012b5eb85b3974f1a959db6e84e55d740)).
   - That behavior is toggled at link time using the `runtime.iscgo` variable.

**Comment by bcmills** [comment]
It looks like the `os` test doesn't have any dependencies that would involve dynamic linking. (Notably, it doesn't end up importing `net`.) I'm not entirely sure whether it is purely statically linked in practice, but you could check by building the test binary with `go test -c os`.

One of these hangs was reported in `os.TestStdPipe` in https://github.com/golang/go/issues/57999#issuecomment-1684522180.

**Comment by ianlancetaylor** [comment]
> I examined a go1.20 executable on netbsd-9 with readelf --dynamic and it certainly looks like it's linked against libpthread:

As @bcmills says, this is going to vary depending on whether the Go program imports any packages that call into C code.  I've verified that on NetBSD the os package tests are running a program that is not dynamically linked.  This is the test used for `TestStdPipe` that @bcmills mentioned.

> If you're calling the _lwp_create syscall directly instead of using the pthread_create symbol, but you're calling any other functions via dynamic symbols, that will get you in trouble and having a newer libpthread.so with the rtld/pthread fix won't help. There is probably lots of other trouble you can get into if you bypass the pthread symbols -- I would have to think more about it.

When dynamically linked, Go programs will call `pthread_create` to create new threads.

That said, there are cases where dynamically linked programs will call the LWP functions.  The functions I see that may be called are `lwp_park`, `lwp_unpark`, `lwp_self`, and `lwp_kill`.

> Can you be more specific about what the difference is?

What we are seeing is that a Go program starts a subprocess, waits for it to complete, and the `wait6(P_PID, pid, WEXITED|WNOWAIT)` call never returns.  If there were a problem such that programs fail to exit, I would expect to also see cases where the build and test run does not complete without reporting any error about waiting for a subprocess.  As far as I know we aren't seeing such a problem.

**Comment by bcmills** [comment]
Hmm. I guess it would be possible to use [`runtime.Stack`](https://pkg.go.dev/runtime#Stack) to detect when the test has been stuck in `wait6` for too long — but then wouldn't we have to start some other new process to prevent the test from deadlocking when it sends itself `SIGSTOP`?

(You're welcome to send a CL to implement that, and I'll review it, but it seems like it would add a lot of extra complexity to an already-complex situation. 😅)

**Comment by bsiegert** [comment]
The flake [above](https://github.com/golang/go/issues/57999#issuecomment-1753807086) is on 10_BETA. So much for my hope that this was fixed in a newer NetBSD version :(

**Comment by bcmills** [comment]
@golang/netbsd, is it possible that this is (somehow) a `kqueue`-related race condition?
#63937 on `darwin` is plausibly `kqueue`-related, and I wonder if they might share a similar root cause.

**Comment by bsiegert** [comment]
> @golang/netbsd, is it possible that this is (somehow) a kqueue-related race condition?
> #63937 on darwin is plausibly kqueue-related, and I wonder if they might share a similar root cause.

Are you thinking of a race in the way Go uses kqueue, or a race in kqueue itself?

**Comment by bcmills** [comment]
I suspect a race in the way Go uses kqueue.
